sdk/iot/hub/samples/methods/src/iot_hub_methods_example.c

```diff
@@ -0,0 +1,430 @@
+    result = az_json_builder_append_token(&json_builder, az_json_token_object_start());
+    az_json_token reported_property_value_token = az_json_token_number((double)return_value);
+    result = az_json_builder_append_object(
+        &json_builder, method_response_name, reported_property_value_token);
+    result = az_json_builder_append_token(&json_builder, az_json_token_object_end());
+
+    response_span = az_json_builder_span_get(&json_builder);
+  }
+  else
+  {
+    response_span = AZ_SPAN_FROM_STR("{}");
+  }
+
+  printf("Status: %u\tPayload:", status);
+  char* payload_char = (char*)az_span_ptr(response_span);
+  for (int32_t i = 0; i < az_span_size(response_span); i++)
+  {
+    putchar(*(payload_char + i));
+  }
+  putchar('\n');
+
+  // Send the response
+  if ((result = MQTTClient_publish(
+           mqtt_client,
+           methods_response_topic,
+           az_span_size(response_span),
+           az_span_ptr(response_span),
+           0,
+           0,
+           NULL))
+      != MQTTCLIENT_SUCCESS)
+  {
+    printf("Failed to publish twin document request, return code %d\n", result);
+    return result;
+  }
+
+  printf("Sent response\n");
+
+  return result;
+}
+
+static int on_received(void* context, char* topicName, int topicLen, MQTTClient_message* message)
+{
+  (void)context;
+
+  if (topicLen == 0)
+  {
+    // The length of the topic if there are one or more NULL characters embedded in topicName,
+    // otherwise topicLen is 0.
+    topicLen = (int)strlen(topicName);
+  }
+
+  az_iot_hub_client_method_request method_request;
+  if (az_iot_hub_client_methods_parse_received_topic(
+          &client, az_span_init((uint8_t*)topicName, topicLen), &method_request)
+      == AZ_OK)
+  {
+    az_result result;
+
+    printf("Direct Method arrived\n");
+    if (az_span_is_content_equal(double_method_span, method_request.name))
+    {
+      // Get payload value
+      az_span payload_span = az_span_init((uint8_t*)message->payload, message->payloadlen);
+      int64_t value_to_double;
+      if ((result = parse_value_to_double(payload_span, &value_to_double)) != AZ_OK)
+      {
+        printf("Unable to parse input value, status %d", result);
+        return result;
+      }
+
+      // Invoke Method
+      int64_t return_value;
+      if ((result = double_method(value_to_double, &return_value)) == AZ_ERROR_ARG)
+      {
+        printf("Unable to invoke double_method, status %d", result);
+        return result;
+      }
+
+      // Build a response
+      if ((result = send_method_response(&method_request, 200, return_value)) != AZ_OK)
+      {
+        printf("Unable to send %d response, status %d", 200, result);
+        return result;
+      }
+    }
+    else
+    {
+      // Unsupported Method
+      if ((result = send_method_response(&method_request, 404, 0xFFFF)) != AZ_OK)
+      {
+        printf("Unable to send %d response, status %d", 404, result);
+        return result;
```

Comment:
>return result; [](start = 8, length = 14)

on_received return bool and indicate to lower layer that you have consumer the message. Now here you are returning non-zero value and not freeing memory of message.

Are you not freeing memory in failure case?
Can you run valgrind on these sample happy path and also while mimicking some failure case (atleast application level.)

Reply:
yes I need to change this to fail a little more gracefully. thanks

Reply:
updated for single return point